### PR TITLE
Number with units parsing [Do not merge][HOLD]

### PR DESCRIPTION
**Pull request by nikitamarysolomanpvt**
Replicated from #190

## Explanation
Part 1: This PR contains Number with input view and it Fix  part #152 .
This is only for finalising Number with units parsing 
## Reference Mock
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/421bb94e-4f53-4943-8ae0-1b6241ba21e6/TM-Q6-Text-Input-w-Units-No-Answer-
## Reference Design Doc
https://docs.google.com/document/d/1QigNdZeMBy-JjHW26QH4qT0UtCyzfvW0xEaStSdY9gM/edit?usp=sharing
## Checklist

- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

### CODE REVIEW

File: data/src/test/assets/api_mocks/currencyunits.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+ {
```
Comment (nikitamarysolomanpvt) [inline data/src/test/assets/api_mocks/currencyunits.json]: This file will be used if Json format is finalised.

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+import org.oppia.app.model.Fraction
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.NumberUnit
+import org.oppia.app.model.NumberWithUnits
+import org.oppia.app.model.UserAnswer
+import org.oppia.app.parser.StringToFractionParser
+import org.oppia.app.player.state.answerhandling.InteractionAnswerHandler
+import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Double.parseDouble
+import java.util.*
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt]: Please don't use wildcard imports: https://developer.android.com/kotlin/style-guide#import_statements.

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    userAnswerBuilder.answer = InteractionObject.newBuilder()
+      .setNumberWithUnits(getNumberWithUnits(text.toString()))
+      .build()
+    userAnswerBuilder.plainAnswer = answerTextString
+
+    return userAnswerBuilder.build()
+  }
+
+  fun getCurrencyUnits(): JSONObject {
+    return JSONObject(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt]: I don't think it makes sense to keep this data in the form of a JSON object. Either we should move it to be an asset that's bundled with the APK (e.g. if we're planning to configure it), move it over to being a response from the server (if we want to remotely change it for already deployed apps), or hardcode it into the application. As-is, we don't need to do any JSON parsing here if we already have all the data in-memory.

For Oppia web, where does this object come from? Is it provided by the server? I think we either need to make a network request to retrieve this file (in which case we need to create a controller for it), or this should be hardcoded directly in the new parsing utility class.

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+import org.oppia.app.model.UserAnswer
+import org.oppia.app.parser.StringToFractionParser
+import org.oppia.app.player.state.answerhandling.InteractionAnswerHandler
+import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Double.parseDouble
+import java.util.*
+import java.util.regex.Pattern
+
+/** The custom EditText class for number input with units interaction view. */
+class NumberWithUnitsInputInteractionView @JvmOverloads constructor(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt:22]: Similar to the other interaction views, please move all of the non view-specific logic to its own file. And, in this case, I suggest adding dedicated tests for that utility. The parsing logic here is much more complicated than the other interactions, and ought to have dedicated tests beyond the ones verifying number with units interaction correctness for this custom view.

Also, please add a link to the Oppia web version of this for reference in the new class.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @nikitamarysolomanpvt. I did a high-level pass on structure, and after you split things up & resolve the embedded JSON object comment I'd like to take a pass on attempting to simplify the parsing logic. After that I'll do the full review.

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt (hunk 2 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt]: Done.

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt (hunk 3 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt]: Inlined the data.

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt (hunk 4 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt:22]: moved to StringToNumberWithUnitsParser file

**Comment by nikitamarysolomanpvt** [comment]
> Thanks @nikitamarysolomanpvt. I did a high-level pass on structure, and after you split things up & resolve the embedded JSON object comment I'd like to take a pass on attempting to simplify the parsing logic. After that I'll do the full review.

Done. PTAL

File: app/src/main/java/org/oppia/app/model/Units.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package org.oppia.app.model
+
+import com.google.gson.annotations.Expose
+import com.google.gson.annotations.SerializedName
+
+class Units(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/model/Units.kt]: This ought to be a data class, and since we aren't interacting with GSON I don't think we need the annotations:

```kotlin
data class Unit(val name: String, val aliases: List<String>, val frontUnits: List<String>, val baseUnit: String? = null)
```

Also note ``val``--data classes should be immutable.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  private lateinit var CURRENCY_UNITS: HashMap<String, Units>
+
+  /** @return index of pattern in s or -1, if not found
+   */
+  fun indexOf(pattern: Pattern, s: String): Int {
+    val matcher = pattern.matcher(s)
+    return if (matcher.find()) matcher.start() else -1
+  }
+
+  fun getCurrencyUnits(): CurrencyUnit {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: Let's initialize this inline above, and we should be using ``listOf()`` and ``mapOf()`` to keep immutability. See the following initial example:

```kotlin
private const val CURRENCY_UNITS = mapOf(
  "dollar" to Unit(name = "dollar", aliases = listOf("$", "dollars", ...), frontUnits = listOf("$")),
  ...
)
```

Note that I intend for these to be defined outside the class since it's actually a constant.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        type = "fraction"
+        fractionObj = StringToFractionParser().parseFractionFromString(value)
+      } else {
+        type = "real"
+        real = Double.parseDouble(value)
+      }
+      if (units != "") {
+        // Checking invalid characters in units.
+        if (units.matches("[^0-9a-z/* ^()₹$-]".toRegex())) {
+          throw Error(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: This error should be removed to be real-time or submit-time error. Ditto for all other errors above.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+          }
+        }
+        if (startsWithCorrectCurrencyUnit == false) {
+          throw  Error("INVALID_CURRENCY")
+        }
+
+        var ind = indexOf(Pattern.compile("[0-9]"), rawInput)
+
+
+        if (ind === -1) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: We should be using ``match`` methods to check presence & extract values rather than doing index-based parsing, similar to what we do in fractions (e.g. https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/app/parser/StringToFractionParser.kt#L83). I think this will simplify a lot of the logic here since I think some of these things can be replaced with patterns (e.g. checking that the value starts with the correct unit).

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        if (ind2 != -1) {
+          value = rawInput.substring(ind, ind2 - ind).trim()
+          units += rawInput.substring(ind2).trim()
+        } else {
+          value = rawInput.substring(ind).trim()
+          units = units.trim()
+        }
+      }
+      // Checking invalid characters in value.
+      if (value.matches("[a - z]".toRegex()) || value.matches("[ *^$₹()#@]/".toRegex())) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: Some of these patterns seem wrong to me--we shouldn't be hardcoding them. They need to be assembled based on the currencies that we've configured in the map. Can you update all relevant patterns to be based on the currency map instead of being hardcoded? Prefer having a single source of truth.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+import org.oppia.app.model.Fraction
+import org.oppia.app.model.NumberUnit
+import org.oppia.app.model.NumberWithUnits
+import org.oppia.app.model.Units
+import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Double
+
+import java.util.regex.Pattern
+
+/** This class contains methods that help to parse string to number, check realtime and submit time errors. */
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: Reminder to add link to Oppia web version of this logic for reference.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @nikitamarysolomanpvt. I think we'll need to go back and forth a few more times. Please address my current comments to further simplify the parsing, then I'll take another pass to see if anything else can be simplified.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt (hunk 6 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: Changed to
 `companion object {
     private val CURRENCY_UNITS = mapOf(
      "dollar" to Units(name = "dollar",...),...)
}`

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt (hunk 9 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: PTAL on https://github.com/oppia/oppia/blob/410fc985080836d348aab101ca4a213a948ff713/core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.ts#L220

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt (hunk 10 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: Done.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt (hunk 7 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: @BenHenning I need to work on final validation and real-time or submit-time error messages once the parsing approach is finalised as communicated already... is that ok?

File: app/src/main/java/org/oppia/app/model/Units.kt (hunk 5 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/model/Units.kt]: Done.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt (hunk 8 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt]: As per your instructions in meeting we are using this to get started.

**Comment by nikitamarysolomanpvt** [comment]
@BenHenning PTAL on parsing approach, I have added basic realtime and submit time validations due to the timeframe and avoid iterations after realtime and submit time validations.
So please suggest validations also and test-cases for the same is in progress. Also error message  content is yet to be finalised.

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 11]
```diff
@@ -0,0 +1,155 @@
+package org.oppia.app.parser
+
+import android.content.Context
+import android.content.res.Configuration
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import org.junit.Before
+import org.junit.Test
```
Comment (nikitamarysolomanpvt) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: @BenHenning PTAL and let me know what other test cases you are expecting

File: app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    typeface = Typeface.DEFAULT
+    showSoftKeyboard(v, context)
+  } else {
+    hint = hintText
+    if (text.isEmpty()) setTypeface(typeface, Typeface.ITALIC)
+    hideSoftKeyboard(v, context)
+  }
+
+  override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
+    if (event.keyCode == KEYCODE_BACK && event.action == ACTION_UP)
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/customview/interaction/NumberWithUnitsInputInteractionView.kt:47]: Nit: please add braces per https://developer.android.com/kotlin/style-guide#braces.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+import org.oppia.app.model.NumberUnit
+import org.oppia.app.model.NumberWithUnits
+import org.oppia.app.model.Unit
+import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Double
+import java.util.regex.Pattern
+
+/**
+ * This class contains methods that help to parse string to number, check realtime and submit time errors.
+ * reference https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.ts#L220.
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt:17]: This reference no longer works (likely because the frontend scripts were recently moved). I suggest using something other than the develop branch (such as a branch hash) to create a permalink and avoid file moves breaking the reference later on.

File: app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+ */
+class StringToNumberWithUnitsParser {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt:19]: For simplification, perhaps just: ``NumberWithUnitsParser``. This also has parity with the fractions parser.

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+import org.oppia.app.model.UserAnswer
+
+@RunWith(AndroidJUnit4::class)
+class StringToNumberWithUnitsParserTest {
+
+  private var context: Context? = null
+  @Before
+  @ExperimentalCoroutinesApi
+  fun setUp() {
+    ApplicationProvider.getApplicationContext<Context>().resources.configuration.orientation =
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: Why is this necessary for a parser test?

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.UserAnswer
+
+@RunWith(AndroidJUnit4::class)
+class StringToNumberWithUnitsParserTest {
+
+  private var context: Context? = null
+  @Before
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: Nit: there should be vertical whitespace between a property list and methods per https://developer.android.com/kotlin/style-guide#vertical.

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+  @ExperimentalCoroutinesApi
+  fun setUp() {
+    ApplicationProvider.getApplicationContext<Context>().resources.configuration.orientation =
+      Configuration.ORIENTATION_LANDSCAPE
+    context = ApplicationProvider.getApplicationContext()
+  }
+
+  fun getPendingAnswer(answerText: String): UserAnswer {
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: This should be private since it's not accessed outside of this class (prefer to minimize visibility).

Also, I suggest ``parseNumberWithUnits`` instead of ``get`` since this isn't actually a traditional fetcher method, it's actually performing parsing logic.

Also, why is this method needed? Why not just call the parser directly?

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    val userAnswerBuilder = UserAnswer.newBuilder()
+    if (answerText.isNotEmpty()) {
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: Prefer to minimize control flow in tests/avoid if-checks.

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+
+  }
+
+  @Test
+  fun testNumberWithUnitsInputInteractionView_withNegativeWholeNumberText_hasCorrectPendingAnswer() {
+    val interactionObject = getPendingAnswer("-9").answer
+    assertThat(interactionObject).isInstanceOf(InteractionObject::class.java)
+    assertThat(interactionObject.objectTypeCase).isEqualTo(InteractionObject.ObjectTypeCase.NUMBER_WITH_UNITS)
+    assertThat(interactionObject.numberWithUnits.real).isLessThan(0.0)
+    assertThat(interactionObject.numberWithUnits.real).isEqualTo(-9.0)
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: Here & everywhere else: real values need to be compared using isWithin() since exact float equality shouldn't be relied upon when the float values are being computed.

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    val numberWithUnits = getPendingAnswer("-10.00 m").answer.numberWithUnits
+    assertThat(numberWithUnits.real).isEqualTo(-10.0)
+    assertThat(numberWithUnits.getUnit(0).unit).isEqualTo("m")
+  }
+
+  @Test
+  fun testNumberWithUnitsInputInteractionView_withInputtedWithSymbols_hasCorrectPendingAnswer() {
+    val numberWithUnits = getPendingAnswer("100.5 Km/m^2").answer.numberWithUnits
+    assertThat(numberWithUnits.real).isEqualTo(100.5)
+    assertThat(numberWithUnits.getUnit(0).unit).isEqualTo("Km/m^2")
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: What about verifying the exponent? Shouldn't this be two units, one with Km exponent 1 and m with exponent -2?

File: app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt (hunk 11 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/parser/FractionInteractionViewModelTest.kt]: Did you port over the tests from Oppia web? This part is essential: unless we are sure the same logical test cases pass as web (e.g. https://github.com/oppia/oppia/blob/develop/core/templates/domain/objects/NumberWithUnitsObjectFactorySpec.ts#L43), we can't attempt to refactor to simplify the parser.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Left a few comments, but sending this back your way @nikitamarysolomanpvt. We need to port all of Oppia web's number with units tests over for the parser before we attempt to simplify it. I think per the current tests that something might be wrong since the units aren't matching what I'd expect, at least per the Oppia web tests I saw: https://github.com/oppia/oppia/blob/develop/core/templates/domain/objects/NumberWithUnitsObjectFactorySpec.ts.

**Comment by BenHenning** [comment]
Since this hasn't seen progress in a while, closing it. We can still use it as a reference, or reopen if we want to continue contributing to this branch.
